src/index.tsx

```diff
@@ -1,23 +1,74 @@
 import * as React from 'react'
 import * as ReactDOM from 'react-dom'
 
-import { FeatureToggles, getBrowserQueryFeatures } from '@paralleldrive/react-feature-toggles'
+import { FeatureToggles, getCurrentActiveFeatures, Feature, isActive } from '@paralleldrive/react-feature-toggles'
+import { Actions } from 'actions'
+import { initialFeatures } from 'config/features'
 import { Provider } from 'react-redux'
 import { Router, Route, browserHistory } from 'react-router'
 import { Layout } from './components/Root'
 import { createPoetStore } from './store'
 
-const { store, pages } = createPoetStore()
+async function init(): Promise<void> {
+  const { store, pages } = await createPoetStore()
+  const routes = pages
+    .map((page: any, index: any) => page.routeHook('' + index))
+    .reduce((a: any, b: any) => a.concat(b), [])
 
-const routes = pages.map((page, index) => page.routeHook('' + index)).reduce((a, b) => a.concat(b), [])
+  function handlerRoutes(store: any, pathname: string): void {
+    const state = store.getState()
+    const { user } = state
+    const omitRoutes: ReadonlyArray<any> = ['/', '/login']
+    const worksNoAuth = [pathname].filter(x => typeof x === 'string' && x.indexOf('/works') > -1)
+    const notNeedOuath = omitRoutes.includes(pathname) || worksNoAuth.length
+    if (['/login', '/login/'].includes(pathname) && user.token !== '') browserHistory.push('/')
 
-ReactDOM.render(
-  <Provider store={store}>
-    <FeatureToggles features={getBrowserQueryFeatures()}>
-      <Router history={browserHistory}>
-        <Route component={Layout}>{routes}</Route>
-      </Router>
-    </FeatureToggles>
-  </Provider>,
-  document.getElementById('app')
-)
+    if (!notNeedOuath && user.token === '') browserHistory.push('/login')
+  }
+  function requireAuth(store: any): (route: any, replace: object) => void {
+    return (route: any, replace: object): void => {
+      const pathname = route.location.pathname
+      store.dispatch(Actions.Router.onEnter(pathname))
+      handlerRoutes(store, pathname)
+    }
+  }
+
+  function onChange(store: any): (route: any, replace: object) => void {
+    return (route: any, replace: any): void => {
+      const pathname = replace.location.pathname
+      store.dispatch(Actions.Router.onChange(pathname))
+      handlerRoutes(store, pathname)
+    }
+  }
+
+  function notFound(route: any, replace: object): void {
+    browserHistory.push('/')
+  }
+
+  ReactDOM.render(
+    <Provider store={store}>
+      <FeatureToggles features={getCurrentActiveFeatures({ initialFeatures })}>
+        <Feature>
+          {({ features }) =>
+            isActive('auth', features) ? (
+              <Router history={browserHistory}>
+                <Route component={Layout} onEnter={requireAuth(store)} onChange={onChange(store)}>
```

Comment:
I've tried it a couple of different ways trying to avoid wrapping the entire Router, but I was not successful in that. 
```tsx
<Router history={browserHistory}>
  <Feature>
    {({ features }) =>
      isActive('auth', features) ? (
        <Route component={Layout} onEnter={requireAuth(store)} onChange={onChange(store)}>
          {routes}
        </Route>
      ) : (
        <Route component={Layout}>{routes}</Route>
      )
    }
  </Feature>
  <Route path="*" onEnter={notFound} />
</Router>
```

```tsx
<Router history={browserHistory}>
  <Feature>
    {({ features }) =>
      isActive('auth', features) ? (
        <Route component={Layout} onEnter={requireAuth(store)} onChange={onChange(store)}>
      ) : (
        <Route component={Layout}>
      )
    }
  </Feature>
    {routes}
  </Route>
  <Route path="*" onEnter={notFound} />
</Router>
```

Reply:
Okidoke, let's forgot about this for now then.

Reply:
Just FYI, In the newer versions of React Router, Route needs to be a direct child of Router or it will not work. Spent a few hours debugging the same issue in a diff project while back :D

The reason being, if I remember correctly from a video I watched, the components now talk to each other directly via a child/parent relationship.